django/core/checks/security/base.py

```diff
@@ -219,6 +224,16 @@ def check_allowed_hosts(app_configs, **kwargs):
     return [] if settings.ALLOWED_HOSTS else [W020]
 
 
+@register(Tags.security, deploy=True)
+def check_type_of_allowed_hosts(app_configs, **kwargs):
+    passed_check = (
+        all(isinstance(element, str) for element in settings.ALLOWED_HOSTS)
+        if isinstance(settings.ALLOWED_HOSTS, (list, tuple))
+        else False
+    )
```

Comment:
Slightly simpler:
```suggestion
    passed_check = (
        isinstance(settings.ALLOWED_HOSTS, (list, tuple))
        and all(isinstance(element, str) for element in settings.ALLOWED_HOSTS)
    )
```

`else False` reminds me of: https://adamj.eu/tech/2020/01/17/simplify-your-ifs-that-return-booleans/

Reply:
Unless I'm mis-reading, both of these are going to cause false positives against `sets` and other more exotic but just as correct datastructures, no?

This is the most salient point of the ticket's comments (emphasis mine):

> The validation basically only needs to check it's an **iterable** of **strings**.

For example, I have no idea if it still works, as I've not touched it in many years, but [I had a package](https://github.com/kezabelle/django-allowedsites) which implemented the small surface area of specific duck-type magic methods Django relies on (`__iter__` et al ... the goalposts may have moved further since) and that worked fine. 

If that's still the case, the messaging:
```
ALLOWED_HOSTS has to be list or tuple of string(s), not string.
```
is not correct, and the _hint_ ought perhaps to suggest making sure it's a list or a tuple, but the warning itself probably shouldn't canonicalise a specific type requirement where there isn't one.

Reply:
The setting is documented as: "A list of strings". Any duck-typing has never been officially supported. If magic is needed, a sublass of `list` can be employed.

Your magic container will fail with this check anyway since checks sometimes run without database connectivity. Imo if you want to implement database-based allowed-hosts protection, you should disable django's with `ALLOWED_HOSTS = ["*"]` and write your own middleware.

Reply:
Hey Adam,
My magic container wasn't the point. It substantially predates system checks even being introduced and as I said, no idea if it even works these days. My argument was that it was too restrictive precisely because it's eschewing ducktyping, regardless of what is documented. I'm struggling to think why a warning for (e.g.) a set should be OK when I'd expect it or a frozen variant to be precisely the right container type for such a thing.
I understand the sentiment and where you're coming from, but the list of strings argument doesn't sway me given the check being introduced literally special cases another sequence datatype just because they're _often_ interchangable and super common.

That said, it's also all just IMHO and if special casing is the path forward to catch the 80% rule then so be it. I just wanted to point out its restrictiveness. It is, after all, only a warning.

Reply:
This is mostly following the status quo for system checks. Django already has many system checks for types, for example of model attributes: https://github.com/django/django/blob/f9cfd346f0222d86bfbac26ef5455806965bae6b/django/db/models/base.py#L1574-L1584

Enforcing and documenting more specific types allows Django to make more assumptions and to change without breaking user code.

Reply:
Maybe we could change the warning message to something slightly more tolerant like the following?
`ALLOWED_HOSTS is expected be list or tuple of string(s), not string.`

@adamchainz thx for above nice change.

Reply:
FWIW I actually got bitten by this updating a project. See ticket-33362